scripts/04-effect-for-beginners.md

```diff
@@ -0,0 +1,1622 @@
+
+---
+
+```ts
+// Effect<never, never, void>
+const program = pipe(
+	// ...
+);
+```
+
+notes:
+now our program has an error channel of never, meaning we can be sure it will never error
+
+---
+
+### Error Handling in Generators
+
+```ts [17-19]
+// Effect<never, never, Pokemon>
+const getPokemon = (id: number) =>
+  Effect.gen(function* (_) {
+    const response = yield* _(
+      Effect.tryPromise({
+        try: () => fetch(`https://pokeapi.co/api/v2/pokemon/${id}`),
+        catch: () => new FetchError(),
+      })
+    );
+    const json = yield* _(
+      Effect.tryPromise({
+        try: () => response.json(),
+        catch: () => new JSONError(),
+      })
+    );
+    return yield* _(parsePokemon(json));
+  }).pipe(
+    Effect.catchAll(() => Effect.succeed({ name: "default", weight: 0 }))
+  );
+```
+
+notes:
+In generators all of the code represents the happy path, so we have to add error handling at the end of the generator. Here's what our the gen version of our new getPokemon function looks like. We create variable for response and json, yielding effects like before. At the end we use the pipe method to pass our catchAll functionality.
+
+---
+
+```ts [9-14]
+// type: Effect<never, never, void>
+const program = Effect.gen(function* (_) {
+  const pokemons = yield* _(
+    Effect.all(getRandomNumberArray(10).map(getPokemon))
+  );
+  console.log(pokemons.map(formatPokemon).join("\n"), "\n");
+  const heaviest = yield* _(calculateHeaviestPokemon(pokemons));
+  console.log(`The heaviest pokemon weighs ${heaviest} hectograms!`);
+}).pipe(
+  Effect.catchTag("SameWeightError", (e) =>
+    Effect.sync(() =>
+      console.log(`Two pokemon have the same weight: ${e.weight}`)
+    )
+  )
+);
+```
+
+notes:
+We can do a similar thing for our program effect, adding our catchTag functionality in the pipe method after the generator.
+
+---
+
+```ts
+Effect.tryPromise({
+      try: () => fetch(`https://pokeapi.co/api/v2/pokemon/${id}`),
+      catch: () => new FetchError(),
+    }),
+```
+
+notes:
+Our pokemon app is hard coded to fetch data through a fetch call to a specific api, but what if we wanted to get the data from a different api, or a database, or maybe even add a cache layer within our application to check before making an api call. To do this traditionally you would have to refactor your code on every change, 
+
+```ts
+const getPokemon = (id: number, client: PokemonClient) =>
+  pipe(
+    client.pokemon.getById(id),
+    // ...
+  )
+```
+
+but what if we could abstract the signature of the data fetching to a dependency that we pass to our effects and could change the implementation of somewhere higher up the chain
+
+---
+
+### Adding Requirements
+(dependency injection)
+
+notes:
+Effect provides this behavior through the Requirements field of the Effect type which we have ignored up until now. Lets look at a simple example
+
+---
+```ts
+type Random = {
+  readonly next: () => Effect.Effect<never, never, number>
```

Comment:
Given `Effect` is lazy a signature like `() => Effect` (function that takes no argument) is almost always a mistake, it can directly be:

```ts
interface Random {
  readonly next: Effect.Effect<never, never, number>
}
```

Reply:
I'd also add a note that it is important for the service shape to be unique to avoid cases of providing one service and ending up erasing another, two patterns can be used:

```ts
interface Random {
  readonly _tag: "Random"
  readonly next: Effect.Effect<never, never, number>
}
```

or

```ts
interface RandomService {
  readonly next: Effect.Effect<never, never, number>
}
interface Random {
  readonly _: unique symbol
}
const Random = Context.Tag<Random, RandomService>("@app/Random")
```

I am progressively leaning towards the second, which can also be wrote as:

```ts
interface Random {
  readonly _: unique symbol
}
const Random = Context.Tag<Random, {
  readonly next: Effect.Effect<never, never, number>
}>("@app/Random")
```

for compactness

Reply:
if its ok with you I think im gonna go for the first option as I think its a bit easier to conceptually grasp for the video